src/mlpack/methods/ann/layer/log_softmax_impl.hpp

```diff
@@ -59,16 +59,21 @@ void LogSoftMax<InputDataType, OutputDataType>::Forward(
 
   maxInput.each_row() += arma::log(arma::sum(output));
   output = input - maxInput;
+
+  if (!deterministic)
+  {
+    y = output;
+  }
 }
 
 template<typename InputDataType, typename OutputDataType>
 template<typename eT>
 void LogSoftMax<InputDataType, OutputDataType>::Backward(
-    const arma::Mat<eT>& input,
+    const arma::Mat<eT>& /* input */,
     const arma::Mat<eT>& gy,
     arma::Mat<eT>& g)
 {
-  g = arma::exp(input) + gy;
+  g = gy - arma::exp(y);
```

Comment:
@zoq If I'm not mistaken, we do pass gradients from higher to lower layers with the negative multiplied?

Reply:
@ShikharJ @zoq: Is there a convention we should follow? The `HardTanH` layer test fails on Jacobian Test when activations are fed to the `Backward` function. And its the only layer that failed JacobianTest when I fed the `output` to `Backward` function in place of `input`. Should we add an argument to the JacobianTest whether we want to feed `input` or `activation` to the `Backward` function?